### PR TITLE
CRM Dynamics integration getLeads UTC date mismatch

### LINKED ISSUE #10506: Mautic CRM bundle uses Non UTC times and appends Z (zulu) which I think is wrong
### Mautic Version

4.0.0

### PHP version

PHP 7.4.22

### What browsers are you seeing the problem on?

Not relevant

### What happened?

sprintf('modifiedon ge %sZ', substr($params['start'], 0, '-6'))

This statement Uses Non UTC dates and creates ZULU (UTC) dates from this in the parameters.

There should be something like: 

```
438                         $given = new DateTime($params['start']);
439                         $given->setTimezone(new DateTimeZone("UTC"));
440                         $oparams['$filter'] = sprintf('modifiedon ge %sZ', $given->format("Y-m-d\TH:i:s"));
```

### How can we reproduce this issue?

Step 1: 
Set up and configure microsoft dynamics

Step 2: 
Set up and configure Mautic
Assumption: Run Mautic in NON UTC timezone

Step 3:
Link some contact fields between Mautic and Micorsoft Dynamics.

Step 4:
Run `bin/console mautic:integration:fetchleads -i Dynamics`
And debug the params passed on to Dynamics (docroot/plugins/MauticCrmBundle/Integration/DynamicsIntegration.php function fetchLeads)

Step 5: 
See that the 
`$oparams['$filter'] = sprintf('modifiedon ge %sZ', gmdate(substr($params['start'], 0, '-6')));`
Gives you an incorrect String and will thus fetch (not all) values from the Dynamics server. If Timezone difference high enough.

FIX suggestion:
I have not yet contributed on the mautic project but I think this is the fix:
`438                         $given = new DateTime($params['date']);
439                         $given->setTimezone(new DateTimeZone("UTC"));
440                         $oparams['$filter'] = sprintf('modifiedon ge %sZ', $given->format("Y-m-d\TH:i:s"));`

I will create the needed branch/PR and try to link it. 

### Relevant log output

_No response_

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by woutersf**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [yes ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes https://github.com/mautic/mautic/issues/10506

#### Summary

This statement Uses Non UTC dates and creates ZULU (UTC) dates from this in the parameters. (It explicitely cuts of the +02) 
`printf('modifiedon ge %sZ', substr($params['start'], 0, '-6'))`
Zulu time: https://www.timeanddate.com/worldclock/timezone/zulu 

The result is that in the countries with (eg +02) different timezones a set of records is fetched that is incomplete.

There should be something like:

```
$given = new DateTime($params['start']);
$given->setTimezone(new DateTimeZone("UTC"));
$oparams['$filter'] = sprintf('modifiedon ge %sZ', $given->format("Y-m-d\TH:i:s"));
```

#### Description:

@nickveenhof    Thanks for the help on my first Mautic bug find!
Reviewers: Please be gentle: This is my first commit 😁 . 

See ISSUE: https://github.com/mautic/mautic/issues/10506

#### Steps to test this PR:

Also in Bug report https://github.com/mautic/mautic/issues/10506

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10507"></a>

### CODE REVIEW

**Comment by RCheesley** [comment]
Thanks for the PR @woutersf - please can you check that the email address you signed your commits with is on your GitHub profile (doesn't have to be the default one, just on the profile and confirmed) - that should then placate the bot :)

**Comment by RCheesley** [comment]
@woutersf it looks like we have a couple of things to tweak.

Code style wise, you need to use [single rather than double quotes](https://github.com/mautic/mautic/pull/10507/checks?check_run_id=3811911208#step:8:31) in this line 

mautic/plugins/MauticCrmBundle/Integration/DynamicsIntegration.php L31

-                    $startDate->setTimezone(new DateTimeZone("UTC"));
+                    $startDate->setTimezone(new DateTimeZone('UTC'));

PHPSTAN is also throwing an error about a class not being found:

https://github.com/mautic/mautic/pull/10507/checks?check_run_id=3811911164#step:8:85

It relates to:

plugins/MauticCrmBundle/Integration/DynamicsIntegration.php        

L508 Error: Instantiated class MauticPlugin\MauticCrmBundle\Integration\DateTime not found.
L509 Error: Instantiated class MauticPlugin\MauticCrmBundle\Integration\DateTimeZone not found.

If we can get those two addressed we should hopefully get the green light! You also need to sign the Contributors Agreement please: https://www.mautic.org/contributor-agreement.

Thanks!

**Comment by woutersf** [comment]
- Commits are now signed.
- Contributor agreement emailed signed (that process might be easier than it is now) I quote my partner: "You jump to all these hoops do help them with a commit?" no hard feelings but a form or word download (or just a form) could help?
- Changed quotes 
- Added \DateTime and \DateTimezone use statements.

**Comment by RCheesley** [comment]
> * no hard feelings but a form or word download (or just a form) could help?

There is a link, right at the top, to do it online. Did you mean that this form was too complex?
screenshot-www mautic org-2021 10 08-07_43_45

> * Commits are now signed.

Can you double check that the email you are using is listed on your GitHub account? There still seems to be a problem which you can see from [this message](https://github.com/mautic/mautic/pull/10507#issuecomment-938064784) (maybe it was the old historical commit though?)

Thanks for addressing the codestyle and other issue, let's get the tests running and check it out! :)

**Comment by woutersf** [comment]
My email address is in the github settings. Not sure what to do next?

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
Great - just needs to get the agreement signed (you can do it online using the link at the top of the page) and then fix the codestyle issue. Then we're good to go! :)

**Comment by woutersf** [comment]
> Great - just needs to get the agreement signed (you can do it online using the link at the top of the page) and then fix the codestyle issue. Then we're good to go! :)

Ah yes, it was in my drafts folder 🤦🏻  - My bad. Email en route...  💌

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by escopecz** [comment]
@woutersf please add a unit test to this change. It cannot be merged without it.

File: plugins/MauticCrmBundle/Integration/DynamicsIntegration.php [hunk 1]
```diff
@@ -479,6 +481,21 @@ public function getLeads($params = [], $query = null, &$executed = null, $result
...
     }
 
+    /**
+     * Convert to UTC date string for API call.
+     *
+     * @param string $paramDate
+     *
+     * @return string
+     */
+    public function getFilterDateUTC($paramDate)
```
Comment (escopecz) [inline plugins/MauticCrmBundle/Integration/DynamicsIntegration.php]: Looking good. Let's use the PHP 7.4 types and save some space :)
```suggestion
    /**
     * Convert to UTC date string for API call.
     */
    public function getFilterDateUTC(string $paramDate): string
```

**Review by escopecz — DISMISSED** [review]
The code change looks good. The test is not optimal as a new public method had to be written for it, but it's way better than no test at all 👍 

_I did not test this, just code reviewed_

**Comment by kuzmany** [comment]
Close favor to M5 version https://github.com/mautic/mautic/pull/13209

**Pull request closed without merging**